File: Cpp/2559_count_vowel_strings_in_ranges.cpp
// https://leetcode.cn/problems/count-vowel-strings-in-ranges/

class Solution {
public:
    vector<int> vowelStrings(vector<string>& words, vector<vector<int>>& queries) {
        int n = words.size();
        vector<int> vowels(26, 0);
        vowels['a' - 'a'] = 1;
        vowels['e' - 'a'] = 1;
        vowels['i' - 'a'] = 1;
        vowels['o' - 'a'] = 1;
        vowels['u' - 'a'] = 1;
        vector<int> valid(n+1, 0);
        for (int i = 0; i < n; ++i) {
            string word = words[i];
            if (vowels[word[0]-'a'] && vowels[word.back()-'a']) {
                valid[i+1] = valid[i] + 1;
            }
            else {
                valid[i+1] = valid[i];
            }
        }

        vector<int> res(queries.size(), 0);
        for (int k = 0; k < queries.size(); ++k) {
            res[k] = valid[queries[k][1]+1] - valid[queries[k][0]];
        }
        return res;
    }
};

File: Python3/2559_count_vowel_strings_in_ranges.py
# https://leetcode.cn/problems/count-vowel-strings-in-ranges/

class Solution:
    def vowelStrings(self, words, queries):
        """
        Inputs:
            words: list[str]
            queries: list[list[int]]
        Outputs:
            res: list[int]
        """
        n = len(words)
        vowels = {"a", "e", "i", "o", "u"}
        valid = [0 for i in range(n+1)]
        for i, word in enumerate(words):
            if word[0] in vowels and word[-1] in vowels:
                valid[i+1] = valid[i] + 1
            else:
                valid[i+1] = valid[i]
        
        res = [0 for k in range(len(queries))]
        for k, query in enumerate(queries):
            res[k] = valid[query[1]+1] - valid[query[0]]

        return res